### DISCUSSION TITLE
Did the ESM only break support with NextJS?

### DISCUSSION THREAD

**Discussion by stephencweiss** [remarkjs/remark#811] [Q&A] [answered]
Hi! I'm relatively new to Remark and NextJS but I'm finding some interesting behavior and trying to trace down what the issue is. 

I have the following code working in my `next` app:

```ts
import remark from "remark";
import html from "remark-html";

export async function markdownToHtml(markdown: string): Promise<string> {
  const result = await remark()
    .use(html)
    .process(markdown);

  return result.toString();
}
```

```json
// package.json
{
  "dependencies": {
    "next": "^11.1.0",
    "react": "17.0.2",
    "react-dom": "17.0.2"
  },
  "devDependencies": {
    "@types/node": "^16.4.2",
    "@types/react": "17.0.15",
    "eslint": "7.31.0",
    "eslint-config-next": "11.0.1",
    "gray-matter": "^4.0.3",
    "remark": "^13.0.0",
    "remark-html": "^13.0.1",
    "remark-parse": "^10.0.0",
    "remark-react": "^9.0.0",
    "remark-validate-links": "^11.0.0",
    "typescript": "4.3.5",
    "unified": "^10.1.0"
  }
}
```

When I add `remark-validate-links`, however, everything breaks ([v11 is where we added esm](https://github.com/remarkjs/remark-validate-links/releases/tag/11.0.0)). 
`remark-html` is v13 (and 14 is where we [added ESM](https://github.com/remarkjs/remark-html/releases/tag/14.0.0)), 

I did upgrade `next` to `11.1` though, which [includes support for esm externals](https://github.com/vercel/next.js/pull/27069). 

So, I guess I'm wondering, have we tested with NextJS and does the new ESM only approach break everything? Or is it just a configuration change that I need to make?

Thanks!

**Comment by ChristianMurphy** [comment, marked as answer]
It is supported.

ESM has been tested with a number of platforms.
Support and workarounds are documented at https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c
ESM can be used in Next even before 11.1 with https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#im-having-problems-with-esm-and-nextjs
After 11.1 it can be supported with https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#gistcomment-3856649

It should be just a configuration change.

**Reply by stephencweiss** [reply]
Thanks so much @ChristianMurphy! Really appreciate the links and it looks like this will be exactly what I need. 🙏

**Reply by Powersource** [reply]
> ESM can be used in Next even before 11.1 with

From reading the link this sounds incorrect? It sounds only supported from 11.1 onwards.

**Reply by ChristianMurphy** [reply]
It can be supported, the link has changed since this was posted, https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c#gistcomment-3760583 has the content that was originally shown.

In general though, consider using Next 12 which supports ESM natively without any configuration or workarounds. https://nextjs.org/blog/next-12#es-modules-support-and-url-imports
